Requestify the computation of the list of memberwise initialized properties

DougGregor opened the pull request:
Fixes rdar://110776763, a case where initialized stored properties introduced by peer macros weren't working alongside declared stored properties.

DougGregor committed:
Requestify the computation of the list of memberwise initialized properties.

Fixes rdar://110776763, a case where initialized stored properties
introduced by peer macros weren't working alongside declared stored
properties.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

xedin commented on lib/Sema/TypeCheckStorage.cpp:
```diff
@@ -354,6 +354,36 @@ InitAccessorPropertiesRequest::evaluate(Evaluator &evaluator,
+    member->visitAuxiliaryDecls([&](Decl *auxDecl) {
+      if (auto auxVar = dyn_cast<VarDecl>(auxDecl))
+        maybeAddProperty(auxVar);
+    });
```
If I understand this correctly there might be a problem with the way this is set up because `init accessors` subsume initialization of some stored properties which means that loop can discover a stored property and add it to results but later discover a property with an `init accessor` (i.e. added by macro expansion) that should actually be used instead. Currently `createImplicitConstructor` establishes mappings between stored properties and their init accessor initializing properties upfront before iterating over the members (assuming that expansions already happened which might not be correct).

xedin replied:
I think we should integrate that to this method and let `createImplicitConstructor` just iterate over the properties instead.

DougGregor replied:
`computeLoweredProperties` forces everything to be expanded that will be expanded. The just-pushed https://github.com/apple/swift/pull/66867/commits/9ea6dcac76134fa488f885273a0288f73433814b does the filtering out of stored properties that are subsumed by computed properties with init accessors, as you're suggesting, so  `createImplicitConstructor` doesn't need special logic.

xedin replied:
Looks good thank you!

DougGregor committed:
Correctly remove properties subsumed by init accessors from memberwise inits

The code to remove these subsumed properties was part of creating the
memberwise initializer declaration, which could lead to
inconsistencies with the newly-introduced request for gathering the
memberwise-initialized properties. Move the code into the computation
of the list of memberwise-initialized properties.

Also update a few more places that were effectively recomputing the
set of memberwise-initialized properties themselves over to the
request.

DougGregor committed:
Reimplement InitAccessorPropertiesRequest on top of MemberwiseInitPropertiesRequest

Simplify the implementation and take advantage of request caching.

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
Make sure we query `hasInitAccessor` before looking for the init accessor.

This is a workaround for the lack of a request for getting accessors.

DougGregor commented:
@swift-ci please smoke test
